apiserver/facades/agent/upgrader/unitupgrader.go

```diff
@@ -125,6 +118,82 @@ func (u *UnitUpgraderAPI) DesiredVersion(ctx context.Context, args params.Entiti
 	return params.VersionResults{Results: result}, nil
 }
 
+// SetTools is responsible for updating a a set of entities reported agent
+// version.
+func (u *UnitUpgraderAPI) SetTools(
+	ctx context.Context,
+	args params.EntitiesVersion,
+) (params.ErrorResults, error) {
+	results := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.AgentTools)),
+	}
+
+	for i, entityVersion := range args.AgentTools {
+		tag, err := names.ParseTag(entityVersion.Tag)
+		if err != nil {
+			results.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+
+		if !u.authorizer.AuthOwner(tag) {
+			results.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+
+		err = u.setEntityToolVersion(ctx, tag, entityVersion)
+		results.Results[i].Error = apiservererrors.ServerError(err)
+	}
+
+	return results, nil
+}
+
+// setEntityToolVersion is responsible for taking a Juju entity identified by
+// tag and setting it's reported agent version in the relevant domain.
+func (u *UnitUpgraderAPI) setEntityToolVersion(
```

Comment:
Get rid of `Tool`

Reply:
This is inline with the context that it is being used. We need to refactor this facade if you want to get rid of Tool.